src/composable/types/twap.ts

```diff
@@ -0,0 +1,244 @@
+    if (!(data.n.gt(1) && data.n.lte(2 ** 32))) throw new Error('InvalidNumParts')
+    if (!(data.t.gt(0) && data.t.lte(365 * 24 * 60 * 60))) throw new Error('InvalidFrequency')
+    if (!data.span.lte(data.t)) throw new Error('InvalidSpan')
+    return true
+  }
+
+  /**
+   * Serialize the TWAP order into it's ABI-encoded form.
+   * @returns {string} The ABI-encoded TWAP order.
+   */
+  serialize(): string {
+    return BaseConditionalOrder.encodeParams(this.leaf)
+  }
+
+  /**
+   * Get the encoded static input for the TWAP order.
+   * @returns {string} The ABI-encoded TWAP order.
+   */
+  encodeStaticInput(): string {
+    return super.encodeStaticInputHelper(TWAP_DATA_ABI, this.staticInput)
+  }
+
+  /**
+   * Deserialize a TWAP order from it's ABI-encoded form.
+   * @param {string} s ABI-encoded TWAP order to deserialize.
+   * @returns A deserialized TWAP order.
+   */
+  static deserialize(s: string): TWAP {
+    return super.deserializeHelper(
+      s,
+      TWAP_ADDRESS,
+      TWAP_DATA_ABI,
+      (o: TWAPData, salt: string) =>
+        new TWAP(
+          TWAP_ADDRESS,
+          salt,
+          {
+            ...o,
+            ...TWAP.partsToTotal(o),
+          },
+          false
+        )
+    )
+  }
+
+  /**
+   * Create a human-readable string representation of the TWAP order.
+   * @param {((address: string, amount: BigNumber) => string) | undefined} tokenFormatter An optional
+   *        function that takes an address and an amount and returns a human-readable string.
+   * @returns {string} A human-readable string representation of the TWAP order.
+   */
+  toString(tokenFormatter?: (address: string, amount: BigNumber) => string): string {
+    const defaultFormatter = (address: string, amount: BigNumber) => `${address}@${amount}`
+    tokenFormatter = tokenFormatter || defaultFormatter
+
+    const { sellAmount, buyAmount } = TWAP.partsToTotal(this.staticInput)
+
+    return `${this.orderType}: Sell total ${tokenFormatter(
+      this.staticInput.sellToken,
+      sellAmount
+    )} for a minimum of ${tokenFormatter(this.staticInput.buyToken, buyAmount)} over ${
+      this.staticInput.n
+    } parts with a spacing of ${this.staticInput.t}s beginning at ${
+      this.staticInput.t0.eq(0) ? 'time of mining' : new Date(Number(this.staticInput.t0) * 1000)
+    }`
+  }
+
+  /**
+   * Transform parameters into a native struct.
+   * @param {TWAPDataParams} params As passed by the consumer of the API.
+   * @returns {TWAPData} A formatted struct as expected by the smart contract.
+   */
+  transformParamsToData(params: TWAPDataParams): TWAPData {
+    return {
+      ...params,
+      ...TWAP.totalToPart(params),
+    }
+  }
+
+  /**
+   * Convert TWAP parts to total amounts.
+   * @param {TWAPData} o The TWAP order.
+   * @returns {object} The total sell amount and total minimum buy amount.
+   */
+  private static partsToTotal(o: TWAPData): {
```

Comment:
Probably it makes sense to remove the `static` modificator and even make it public.
This method at least will be useful in CowSwap

Reply:
I think it would be helpful to understand CoW Swap's use of this. Nominally, the way the `TWAP` class is designed is that it takes a total `sellAmount`, and total `minBuyAmount`, and from this, with the given number of discrete orders, it will do all the math, without it being the concern of the caller.

In this case, the caller should only ever care about the total value in / out of the TWAP.

On area where I can see it being useful is to have an "order generator", so given an implementation of `BaseConditionalOrder`, such as `TWAP`, the API provides a `get discreteOrders: GPv2Order.Data` the returns all the discrete orders belong to a generator type conditional order?

Reply:
Discrete order generator tracked in #144